### ISSUE TITLE
BHTree doesn't commit particles after adding them

### ISSUE THREAD

**Issue by webbjj** [amusecode/amuse#518]
**Describe the bug**
I have been trying to run a very simple star cluster evolution script, where the cluster evolves in an external tidal field while undergoing stellar evolution. 

My typical approach is:
galaxy_code=MWpotentialBovy2015()
cluster_code=BHTree()
--> cluster_code.add_particles
stellar_code=SSE()
-->stellar_code.add_particles
setup bridge
evolve

However this does not seem to work and produces the error: `amuse.support.exceptions.AmuseException: Error when calling 'set_mass' of a 'BHTree', errorcode is -1`

However if I use PH4 instead of BHTree, the code runs fine. Alternatively, while still using BHTree, if I wait until after SSE is called once to add particles to cluster_code, then the code runs fine as well. So I am guessing there is some issue with the BHTree interface?

My script is copied below. This script doesn't work unless the lines underneath where I write (DEBUG: APPARENTLY STARS CANNOT BE ADDED BELOW) are commented out and the lines underneath #DEBUG: WORKS WHEN STARS ARE ADDED BELOW are uncommented.

Any help would be most appreciated.

Jeremy

CODE:
```python
import matplotlib
matplotlib.use('Agg')

from amuse.lab import *
from amuse.couple import bridge
from amuse.support.literature import LiteratureReferencesMixIn
from amuse.datamodel import Particles
from amuse.community.sse.interface import SSE
from amuse.ext.galactic_potentials import *

from galpy import potential
import numpy as np

import matplotlib.pyplot as plt

def evolve_cluster_in_galaxy(N,Rcluster,Rinit,Vinit, galaxy_code, dt, dtout, tend, epsilon):

    #Setup cluster
    masses = new_kroupa_mass_distribution(N)
    Mcluster=masses.sum()

    #Create star cluster with origin at 0,0,0 and no net velocity
    converter=nbody_system.nbody_to_si(Mcluster,Rcluster)
    stars=new_plummer_sphere(N,converter)

    #Place star cluster in Galactocentric position
    stars.x += Rinit[0]
    stars.y += Rinit[1]
    stars.z += Rinit[2]

    stars.vx += Vinit[0]
    stars.vy += Vinit[1]
    stars.vz += Vinit[2]

    stars.mass=masses

    stars.scale_to_standard(convert_nbody=converter, smoothing_length_squared = epsilon**2)

    #Gravity code
    cluster_code=BHTree(converter,number_of_workers=1)     #Change number of workers depending on CPUS available
    cluster_code.parameters.epsilon_squared = epsilon**2
    cluster_code.parameters.opening_angle=0.6
    cluster_code.parameters.timestep=dt
    #DEBUG: APPARENTLY STARS CANNOT BE ADDED BELOW
    cluster_code.particles.add_particles(stars)

    #Setup channels between stars particle dataset and the cluster code
    channel_from_stars_to_cluster_code=stars.new_channel_to(cluster_code.particles, attributes=["mass", "x", "y", "z", "vx", "vy", "vz"])    
    channel_from_cluster_code_to_stars=cluster_code.particles.new_channel_to(stars, attributes=["mass", "x", "y", "z", "vx", "vy", "vz"])

    #Stellar evolution code
    stellar_evolution=SSE()
    stellar_evolution.particles.add_particle(stars)

    #Setup channels between stars particle dataset and stellar evolution code
    channel_from_stars_to_stellar_evolution=stars.new_channel_to(stellar_evolution.particles, attributes=["mass"])    
    channel_from_stellar_evolution_to_stars=stellar_evolution.particles.new_channel_to(stars, attributes=["mass"])

    #Setup gravity bridge
    gravity=bridge.Bridge(use_threading=False)
    #stars in cluster_code depend on gravity from galaxy_code
    gravity.add_system(cluster_code, (galaxy_code,))
    #galaxy_code still needs to be added to system so it evolves with time
    gravity.add_system(galaxy_code,)
    #Set how often to update external potential
    gravity.timestep = cluster_code.parameters.timestep/2.

    time=0.0 | tend.unit
    first=True
    while time<tend:
        print(time.value_in(units.Myr),tend.value_in(units.Myr),stars.mass.sum().value_in(units.MSun))

        stellar_evolution.evolve_model(time+dt/2.)
        channel_from_stellar_evolution_to_stars.copy_attributes(["mass"])
        #DEBUG: APPARENTLY STARS CANNOT BE ADDED BELOW
        channel_from_stars_to_cluster_code.copy_attributes(["mass"])

        #DEBUG: WORKS WHEN STARS ARE ADDED BELOW
        #if first:
        #    cluster_code.particles.add_particles(stars)
         #   first=False
        #else:
        #    channel_from_stars_to_cluster_code.copy_attributes(["mass"])

        gravity.evolve_model(time+dt)
        channel_from_cluster_code_to_stars.copy()

        stellar_evolution.evolve_model(time+dt)
        channel_from_stellar_evolution_to_stars.copy_attributes(["mass"])
        channel_from_stars_to_cluster_code.copy_attributes(["mass"])
       
        time = gravity.model_time

    #Copy back to stars for final dataset
    channel_from_cluster_code_to_stars.copy()
    gravity.stop()

    return stars

if __name__ == "__main__":
    #Set initial cluster parameters
    N=1000
    Rcluster = 10. | units.parsec
    Rinit=[10.,0.,0.] | units.kpc
    Vinit=[0.,220.,0.] | units.km/units.s
    epsilon=0.75 | units.parsec

    #Setup star cluster simulation
    #Simulation end time
    tend = 1000.0 | units.Myr
    #Frequency of data output
    dtout=5.0 | units.Myr
    #Frequency of star cluster gravity calculation
    dt = 1.0 | units.Myr

    #Set Galactic Potential 
    galaxy_code = MWpotentialBovy2015()

    #Evolve star cluster
    stars = evolve_cluster_in_galaxy(N,Rcluster,Rinit,Vinit,galaxy_code, dt, dtout, tend, epsilon)

    #Plot final snapshot
    plt.plot(stars.x.value_in(units.kpc),stars.y.value_in(units.kpc),'.')
    plt.xlabel('X (kpc)')
    plt.ylabel('Y (kpc)')
    plt.savefig('test_cluster.png')
```

**Comment by rieder** [comment]
I find an error in the following line:
```python
stellar_evolution.particles.add_particle(stars)
```

this should be `stellar_evolution.particles.add_particles(stars)`

I'm not sure if this will solve the problem though.

**Comment by webbjj** [comment]
Thanks for the thorough read Steven. So the typo originated from lines 103 and 147 in the gravity_stellar_comparison.py code. I just copied from there. Surprisingly it doesn't seem to make a difference. You would think missing the s would throw an error.

So even with the typo fix, the code does the same thing and gives the same error if I do cluster_code.add_particles(stars) before SSE is called for the first time as opposed to after.

One other thing that may help track down the error. If I do everything the proper way, that is add stars to cluster_code when its first initialized and add stars to stellar_evolution when its first initialized and then do the stellar_evolution --> bridge --> stellar_evolution, I can get it to work if there is a simple "Etot_init = cluster_code.kinetic_energy + cluster_code.potential_energy" right after cluster_code.add_particles(stars). The exact code is below. See where I commented #DEBUG. This code doesn't throw the "amuse.support.exceptions.AmuseException: Error when calling 'set_mass' of a 'BHTree', errorcode is -1" error, but does if the Etot_init call isn't made.

I know the issue is not that this line of code is essentially. But I am guessing the call does something to the stars particle set that allows it to be input into BHTree? Similarly, something happens to the masses when SSE is called that allows them to be readable by BHTree? But Ph4 does not have this issue? 

Jeremy




CODE:

import matplotlib
matplotlib.use('Agg')

from amuse.lab import *
from amuse.couple import bridge
from amuse.support.literature import LiteratureReferencesMixIn
from amuse.datamodel import Particles
from amuse.community.sse.interface import SSE
from amuse.ext.galactic_potentials import *

from galpy import potential
import numpy as np

import matplotlib.pyplot as plt

def evolve_cluster_in_galaxy(N,Rcluster,Rinit,Vinit, galaxy_code, dt, dtout, tend, epsilon):

    #Setup cluster
    masses = new_kroupa_mass_distribution(N)
    Mcluster=masses.sum()

    #Create star cluster with origin at 0,0,0 and no net velocity
    converter=nbody_system.nbody_to_si(Mcluster,Rcluster)
    stars=new_plummer_sphere(N,converter)

    #Place star cluster in Galactocentric position
    stars.x += Rinit[0]
    stars.y += Rinit[1]
    stars.z += Rinit[2]

    stars.vx += Vinit[0]
    stars.vy += Vinit[1]
    stars.vz += Vinit[2]

    stars.mass=masses

    stars.scale_to_standard(convert_nbody=converter, smoothing_length_squared = epsilon**2)

    #Gravity code
    cluster_code=BHTree(converter,number_of_workers=1)     #Change number of workers depending on CPUS available
    cluster_code.parameters.epsilon_squared = epsilon**2
    cluster_code.parameters.opening_angle=0.6
    cluster_code.parameters.timestep=dt
    cluster_code.particles.add_particles(stars)

    #DEBUG: Apparently all is well if the below line is included:
    Etot_init = cluster_code.kinetic_energy + cluster_code.potential_energy

    #Setup channels between stars particle dataset and the cluster code
    channel_from_stars_to_cluster_code=stars.new_channel_to(cluster_code.particles, attributes=["mass", "x", "y", "z", "vx", "vy", "vz"])    
    channel_from_cluster_code_to_stars=cluster_code.particles.new_channel_to(stars, attributes=["mass", "x", "y", "z", "vx", "vy", "vz"])

    #Stellar evolution code
    stellar_evolution=SSE()
    stellar_evolution.particles.add_particles(stars)

    #Setup channels between stars particle dataset and stellar evolution code
    channel_from_stars_to_stellar_evolution=stars.new_channel_to(stellar_evolution.particles, attributes=["mass"])    
    channel_from_stellar_evolution_to_stars=stellar_evolution.particles.new_channel_to(stars, attributes=["mass"])

    #Setup gravity bridge
    gravity=bridge.Bridge(use_threading=False)
    #stars in cluster_code depend on gravity from galaxy_code
    gravity.add_system(cluster_code, (galaxy_code,))
    #galaxy_code still needs to be added to system so it evolves with time
    gravity.add_system(galaxy_code,)
    #Set how often to update external potential
    gravity.timestep = cluster_code.parameters.timestep/2.

    time=0.0 | tend.unit
    while time<tend:
        print(time.value_in(units.Myr),tend.value_in(units.Myr),stars.mass.sum().value_in(units.MSun))

        stellar_evolution.evolve_model(time+dt/2.)
        channel_from_stellar_evolution_to_stars.copy_attributes(["mass"])
        channel_from_stars_to_cluster_code.copy_attributes(["mass"])

        gravity.evolve_model(time+dt)
        channel_from_cluster_code_to_stars.copy()

        stellar_evolution.evolve_model(time+dt)
        channel_from_stellar_evolution_to_stars.copy_attributes(["mass"])
        channel_from_stars_to_cluster_code.copy_attributes(["mass"])
       
        time = gravity.model_time

    #Copy back to stars for final dataset
    channel_from_cluster_code_to_stars.copy()
    gravity.stop()

    return stars

if __name__ == "__main__":
    #Set initial cluster parameters
    N=1000
    Rcluster = 10. | units.parsec
    Rinit=[10.,0.,0.] | units.kpc
    Vinit=[0.,220.,0.] | units.km/units.s
    epsilon=0.75 | units.parsec

    #Setup star cluster simulation
    #Simulation end time
    tend = 1000.0 | units.Myr
    #Frequency of data output
    dtout=5.0 | units.Myr
    #Frequency of star cluster gravity calculation
    dt = 1.0 | units.Myr

    #Set Galactic Potential 
    galaxy_code = MWpotentialBovy2015()

    #Evolve star cluster
    stars = evolve_cluster_in_galaxy(N,Rcluster,Rinit,Vinit,galaxy_code, dt, dtout, tend, epsilon)

    #Plot final snapshot
    plt.plot(stars.x.value_in(units.kpc),stars.y.value_in(units.kpc),'.')
    plt.xlabel('X (kpc)')
    plt.ylabel('Y (kpc)')
    plt.savefig('test_cluster.png')

**Comment by rieder** [comment]
Line 224 in BHTree's interface.cc (`if (i >= 0 && i < bhtcs.n) {`) causes the failure, since bhtrc.n is not yet defined when commit_particles has not yet been called.
Solution:
- call commit_particles just in time (e.g. when setting particle attributes) automatically
- initialise the bhtree system when adding the first particle
- ?
@ipelupessy what do you think is best?

**Comment by ipelupessy** [comment]
@rieder euh, I don't understand what happens here yet, can you send me a summary and maybe minimmal script?

**Comment by rieder** [comment]
This minimal script illustrates the problem. A `g.commit_particles()` call after add_particles solves the error, but we shouldn't have to do that manually.
```python
Python 3.8.0 (default, Nov  3 2019, 10:55:54) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.10.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from amuse.lab import * 
   ...: conv = nbody_system.nbody_to_si(1|units.MSun,1|units.parsec) 
   ...: s = new_plummer_model(100, conv) 
   ...: g = BHTree(conv) 
   ...: g.particles.add_particles(s) 
   ...: g.particles.mass = 0.1 | units.MSun                                                                                                                              
---------------------------------------------------------------------------
AmuseException                            Traceback (most recent call last)
<ipython-input-1-ecfa4410cbe6> in <module>
      4 g = BHTree(conv)
      5 g.particles.add_particles(s)
----> 6 g.particles.mass = 0.1 | units.MSun

~/Env/Amuse13/lib/python3.8/site-packages/amuse/datamodel/particles.py in __setattr__(self, name_of_the_attribute, value)
   1105             self._derived_attributes[name_of_the_attribute].set_values_for_entities(self, value)
   1106         else:
-> 1107             self.set_values_in_store(None, [name_of_the_attribute], [self._convert_from_entities_or_quantities(value)])
   1108 
   1109     def get_containing_set(self):

~/Env/Amuse13/lib/python3.8/site-packages/amuse/datamodel/particles.py in set_values_in_store(self, indices, attributes, values)
   1367 
   1368     def set_values_in_store(self, indices, attributes, values):
-> 1369         self._private.attribute_storage.set_values_in_store(indices, attributes, values)
   1370 
   1371     def set_values_in_store_async(self, indices, attributes, values):

~/Env/Amuse13/lib/python3.8/site-packages/amuse/datamodel/incode_storage.py in set_values_in_store(self, indices_in_the_code, attributes, values)
    971 
    972         for setter in self.select_setters_for(attributes):
--> 973             setter.set_attribute_values(self, attributes, values, indices_in_the_code)
    974 
    975     def set_values_in_store_async(self, indices_in_the_code, attributes, values):

~/Env/Amuse13/lib/python3.8/site-packages/amuse/datamodel/incode_storage.py in set_attribute_values(self, storage, attributes, values, *indices)
    329         list_arguments.extend(list_args)
    330         keyword_args.update(storage.extra_keyword_arguments_for_getters_and_setters)
--> 331         self.method(*list_arguments, **keyword_args)
    332 
    333 

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/methods.py in __call__(self, *list_arguments, **keyword_arguments)
    165         object = self.precall()
    166         list_arguments, keyword_arguments = self.convert_arguments(list_arguments, keyword_arguments)
--> 167         result = self.method(*list_arguments, **keyword_arguments)
    168 
    169         result = self.convert_result(result)

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/methods.py in __call__(self, *list_arguments, **keyword_arguments)
    167         result = self.method(*list_arguments, **keyword_arguments)
    168 
--> 169         result = self.convert_result(result)
    170 
    171         self.postcall(object)

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/methods.py in convert_result(self, result)
    206 
    207     def convert_result(self, result):
--> 208         return self.definition.convert_result(self, result)
    209 
    210     def convert_result_index(self):

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/interface.py in convert_result(self, method, result)
    629 
    630     def convert_result(self, method, result):
--> 631         return self.handle_return_value(method, result)
    632 
    633     # this function tries to determine the size of results from definition

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/interface.py in handle_as_unit(self, method, return_value)
    565     def handle_as_unit(self, method, return_value):
    566         if not self.is_return_units_iterable:
--> 567             return self.return_units.convert_result_value(method, self, return_value)
    568         else:
    569             if not hasattr(return_value, '__iter__'):

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/interface.py in convert_result_value(self, method, definition, errorcode)
     69         if hasattr(errorcode, '__iter__'):
     70             for x in errorcode:
---> 71                 definition.handle_errorcode(x)
     72         else:
     73             definition.handle_errorcode(errorcode)

~/Env/Amuse13/lib/python3.8/site-packages/amuse/support/interface.py in handle_errorcode(self, errorcode)
    549             raise exceptions.AmuseException("Error when calling '{0}' of a '{1}', errorcode is {2}, error is '{3}'".format(self.name, type(self.wrapped_object).__name__, errorcode,  self.wrapped_object.errorcodes[errorcode]), errorcode)
    550         elif errorcode < 0:
--> 551             raise exceptions.AmuseException("Error when calling '{0}' of a '{1}', errorcode is {2}".format(self.name, type(self.wrapped_object).__name__, errorcode), errorcode)
    552         else:
    553             return errorcode

AmuseException: Error when calling 'set_mass' of a 'BHTree', errorcode is -1
```

**Comment by ipelupessy** [comment]
ok, so the set_mass needs the particles to be committed?

**Comment by ipelupessy** [comment]
There is nothing in define_state in interface/gd.py about set_mass, set_position etc (there is about get_mass)..I think there should..

**Comment by ipelupessy** [comment]
in fact if you add 
```print(g.particles[0].mass)```
just before the setting of the mass, it runs...

**Comment by rieder** [comment]
yes

**Comment by rieder** [comment]
that too :). But it should run without needing to do that.

**Comment by rieder** [comment]
> There is nothing in define_state in interface/gd.py about set_mass, set_position etc (there is about get_mass)..I think there should..

I think that's the "real" solution indeed.

**Comment by ipelupessy** [comment]
PR #551

**Comment by ipelupessy** [comment]
it needs to be tested with bhtree and other gd codes...

**Comment by rieder** [comment]
Fixed by #551
